### PR TITLE
Cache iast metrics

### PR DESCRIPTION
**Pull request by iunanua**
### What does this PR do?
Reduces the number of times certain IAST telemetry metrics are created by caching and reusing them.

### Motivation
The [`getId`](https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/telemetry/metrics.js#L5) method from the `telemetry/metrics.js` module is a bit expensive and in cases such as in the `executed.[sink|source]` iast metrics `getId` method was called a ton of times. So the runtime performance can be improved caching the metrics once they are created.

### Plugin Checklist

- [ ] Unit tests.
- [ ] TypeScript [definitions][1].
- [ ] TypeScript [tests][2].
- [ ] API [documentation][3].
- [ ] CircleCI [jobs/workflows][4].
- [ ] Plugin is [exported][5].

[1]: https://github.com/DataDog/dd-trace-js/blob/master/index.d.ts
[2]: https://github.com/DataDog/dd-trace-js/blob/master/docs/test.ts
[3]: https://github.com/DataDog/documentation/blob/master/content/en/tracing/trace_collection/library_config/nodejs.md
[4]: https://github.com/DataDog/dd-trace-js/blob/master/.circleci/config.yml
[5]: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/plugins/index.js

### Security 
Datadog employees:
- [ ] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [ ] This PR doesn't touch any of that.

Unsure? Have a question? Request a review!

[APPSEC-43070]

[APPSEC-43070]: https://datadoghq.atlassian.net/browse/APPSEC-43070?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js [hunk 1]
```diff
@@ -19,37 +19,54 @@ const TagKey = {
...
-  getTag (tagValue) {
-    return tagValue ? { [this.tagKey]: tagValue } : undefined
+  add (context, value, tags) {
+    const namespace = getNamespace(this.scope, context)
+    namespace.getMetric(this.name, tags).inc(value)
   }
 
-  addValue (value, tagValue, context) {
-    this.getNamespace(context)
-      .count(this.name, this.getTag(tagValue))
```
[the remark this answers is not in the document]
Comment (iunanua) [reply packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js:39]: it has been moved to `IastNamespace.getMetric` in order the namespace can manage the cache.

File: packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js [hunk 2]
```diff
@@ -19,37 +19,54 @@ const TagKey = {
...
+  constructor (name, scope) {
+    super(name, scope)
 
-  add (value, tagValue, context) {
-    if (Array.isArray(tagValue)) {
-      tagValue.forEach(tag => this.addValue(value, tag, context))
-    } else {
-      this.addValue(value, tagValue, context)
-    }
+    this.tags = []
```
[the remark this answers is not in the document]
Comment (iunanua) [reply packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js:60]: the tags are used as the cache key so each IastMetric to be cacheable needs its own tags property

File: packages/dd-trace/src/appsec/iast/telemetry/namespaces.js [hunk 3]
```diff
@@ -40,11 +40,14 @@ function finalizeRequestNamespace (context, rootSpan) {
...
-    globalNamespace
-      .count(metric.metric, getTagsObject(metric.tags))
-      .inc(point[1])
-  }))
+  metrics.forEach(metric => {
+    const { metric: metricName, type, tags, points } = metric
+
+    if (points?.length) {
+      const gMetric = globalNamespace[type](metricName, getTagsObject(tags))
+      points.forEach(point => gMetric.inc(point[1]))
```
Comment (iunanua) [inline packages/dd-trace/src/appsec/iast/telemetry/namespaces.js:48]: check if metric is 'count' type

File: packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js [hunk 4]
```diff
@@ -61,21 +70,21 @@ function getInstrumentedMetric (tagKey) {
...
 
 // DEBUG using metrics
-const EXECUTED_PROPAGATION = new IastMetric('executed.propagation', Scope.REQUEST)
-const EXECUTED_TAINTED = new IastMetric('executed.tainted', Scope.REQUEST)
+const EXECUTED_PROPAGATION = new NoTaggedIastMetric('executed.propagation', Scope.REQUEST)
+const EXECUTED_TAINTED = new NoTaggedIastMetric('executed.tainted', Scope.REQUEST)
 
 // DEBUG using distribution endpoint
-const INSTRUMENTATION_TIME = new IastMetric('instrumentation.time', Scope.GLOBAL)
+const INSTRUMENTATION_TIME = new NoTaggedIastMetric('instrumentation.time', Scope.GLOBAL)
```
Comment (uurien) [inline packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js:87]: Is this metric used somewhere?

Comment (iunanua) [reply packages/dd-trace/src/appsec/iast/telemetry/iast-metric.js]: removed

**Review by uurien — APPROVED** [review]

**Pull request merged**